src/lib.rs

```diff
@@ -257,12 +257,13 @@ impl<P: AsRef<[u8]>, T: Transitions> Automaton<P> for AcAutomaton<P, T> {
     #[inline]
     fn next_state(&self, mut si: StateIdx, b: u8) -> StateIdx {
         loop {
-            let maybe_si = self.states[si as usize].goto(b);
+            let state = &self.states[si as usize];
+            let maybe_si = state.goto(b);
             if maybe_si != FAIL_STATE {
                 si = maybe_si;
                 break;
             } else {
-                si = self.states[si as usize].fail;
```

Comment:
It is rather odd that this bounds check does not get optimized away though (or perhaps it does get optimized away and the code sent to llvm is just simpler after letting it do other optimizations).

Reply:
Interesting!